flink-tests/src/test/java/org/apache/flink/test/state/operator/restore/AbstractOperatorRestoreTestBase.java

```diff
@@ -104,175 +91,97 @@ public static void beforeClass() {
+
 		URL savepointResource = AbstractOperatorRestoreTestBase.class.getClassLoader().getResource("operatorstate/" + getMigrationSavepointName());
 		if (savepointResource == null) {
 			throw new IllegalArgumentException("Savepoint file does not exist.");
 		}
 		JobGraph jobToMigrate = createJobGraph(ExecutionMode.MIGRATE);
 		jobToMigrate.setSavepointRestoreSettings(SavepointRestoreSettings.forPath(savepointResource.getFile()));
 
-		Object msg;
-		Object result;
+		assertNotNull(jobToMigrate.getJobID());
 
-		// Submit job graph
-		msg = new JobManagerMessages.SubmitJob(jobToMigrate, ListeningBehaviour.DETACHED);
-		result = Await.result(jobManager.ask(msg, timeout), timeout);
+		clusterClient.submitJob(jobToMigrate, classLoader);
 
-		if (result instanceof JobManagerMessages.JobResultFailure) {
-			JobManagerMessages.JobResultFailure failure = (JobManagerMessages.JobResultFailure) result;
-			throw new Exception(failure.cause());
-		}
-		Assert.assertSame(JobManagerMessages.JobSubmitSuccess.class, result.getClass());
+		CompletableFuture<JobStatus> jobStatusFuture = FutureUtils.retrySuccesfulWithDelay(
+			() -> clusterClient.getJobStatus(jobToMigrate.getJobID()),
+			deadline.timeLeft().toMillis() / 50,
+			Time.milliseconds(50),
+			(jobStatus) -> jobStatus.equals(JobStatus.RUNNING),
+			TestingUtils.defaultScheduledExecutor());
 
-		// Wait for all tasks to be running
-		msg = new TestingJobManagerMessages.WaitForAllVerticesToBeRunning(jobToMigrate.getJobID());
-		Await.result(jobManager.ask(msg, timeout), timeout);
+		assertEquals(JobStatus.RUNNING, jobStatusFuture.get(deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS));
 
 		// Trigger savepoint
 		File targetDirectory = tmpFolder.newFolder();
-		msg = new JobManagerMessages.CancelJobWithSavepoint(jobToMigrate.getJobID(), targetDirectory.getAbsolutePath());
+		String savepointPath = null;
 
 		// FLINK-6918: Retry cancel with savepoint message in case that StreamTasks were not running
 		// TODO: The retry logic should be removed once the StreamTask lifecycle has been fixed (see FLINK-4714)
-		boolean retry = true;
-		for (int i = 0; retry && i < 10; i++) {
-			Future<Object> future = jobManager.ask(msg, timeout);
-			result = Await.result(future, timeout);
-
-			if (result instanceof JobManagerMessages.CancellationFailure) {
-				Thread.sleep(50L);
-			} else {
-				retry = false;
+		while (deadline.hasTimeLeft() && savepointPath == null) {
+			try {
+				savepointPath = clusterClient.cancelWithSavepoint(
+					jobToMigrate.getJobID(),
+					targetDirectory.getAbsolutePath());
+			} catch (Exception e) {
+				if (!e.toString().matches(".* savepoint for the job .* failed.*")) {
+					throw e;
+				}
 			}
 		}
 
-		if (result instanceof JobManagerMessages.CancellationFailure) {
-			JobManagerMessages.CancellationFailure failure = (JobManagerMessages.CancellationFailure) result;
-			throw new Exception(failure.cause());
-		}
+		assertNotNull(savepointPath);
```

Comment:
add error message that the savepoint could not be taken

Reply:
fixing